drivers/dma/dma_silabs_siwx917.c

```diff
@@ -0,0 +1,634 @@
+	}
+	// Store the UDMA handle
+	siwx917_udma_handle[dma_init.dma_number] = dma_init.udma_handle;
+
+	// Connect the DMA interrupt
+	cfg->irq_configure();
+
+	// Enable UDMA instance
+	if (sl_si91x_dma_enable(dma_init.dma_number) != SL_STATUS_OK) {
+		return -EBUSY;
+	}
+	return status;
+}
+
+static void dma_siwx917_isr(const struct device *dev)
+{
+	// Fetch DMA config and data structures from device.
+	const struct dma_siwx917_config *cfg = dev->config;
+	struct dma_siwx917_data *data = dev->data;
+	uint32_t irq_number;
+	UDMA_Channel_Info channel_info[32] = {0};
+	if (cfg->reg == UDMA0) {
+		irq_number = UDMA0_IRQn;
+	} else {
+		irq_number = UDMA0_IRQn;
+	}
+	// Disable IRQ
+	irq_disable(irq_number);
+	uint8_t channel;
+	uint32_t int_status;
+	uint8_t soft_trig_flag = 0;
+	uint8_t transfer_complete = 0;
+	int_status = cfg->reg->UDMA_DONE_STATUS_REG; // Read the interrupt status
+	// Initialize UDMA Resources
+	UDMA_RESOURCES UDMA_Siwx917_Resources = {
+		cfg->reg, irq_number,
+		(RSI_UDMA_DESC_T *)cfg->reg->CTRL_BASE_PTR // SRAM base address
+	};
+
+	// Identify the interrupt channel
+	for (channel = 0; channel < cfg->channels; channel++) {
+		if (int_status & (1U << channel)) {
+			// Copy channel info to channel_info[].
+			channel_info[channel].Cnt = data[channel].siwx917_udma_channel_info.Cnt;
+			channel_info[channel].DestAddr =
+				data[channel].siwx917_udma_channel_info.DestAddr;
+			channel_info[channel].Size = data[channel].siwx917_udma_channel_info.Size;
+			channel_info[channel].SrcAddr =
+				data[channel].siwx917_udma_channel_info.SrcAddr;
+			if (channel_info[channel].Cnt != channel_info[channel].Size) {
+				// Check if the transfer type is memory-memory
+				if ((UDMA_Siwx917_Resources.desc[channel]
+					     .vsUDMAChaConfigData1.srcInc != UDMA_SRC_INC_NONE) &&
+				    (UDMA_Siwx917_Resources.desc[channel]
+					     .vsUDMAChaConfigData1.dstInc != UDMA_DST_INC_NONE)) {
+					// Need to apply a software trigger later
+					soft_trig_flag = 1;
+				}
+				if ((channel_info[channel].Size - channel_info[channel].Cnt) >
+				    DMA_MAX_TRANSFER_COUNT) {
+					// Update dev->data with latest transfer count.
+					data[channel].siwx917_udma_channel_info.Cnt +=
+						DMA_MAX_TRANSFER_COUNT;
+				} else {
+					data[channel].siwx917_udma_channel_info.Cnt +=
+						(channel_info[channel].Size -
+						 channel_info[channel].Cnt);
+				}
+			} else {
+				transfer_complete = 1;
+			}
+			break;
+		}
+	}
+	// Call UDMA ROM IRQ handler.
+	ROMAPI_UDMA_WRAPPER_API->uDMAx_IRQHandler(&UDMA_Siwx917_Resources,
+						  UDMA_Siwx917_Resources.desc, channel_info);
+
+	if (soft_trig_flag) {
+		// Set the software trigger bit for starting next transfer
+		cfg->reg->CHNL_SW_REQUEST |= (1U << channel);
+	}
+	if ((data[channel].dma_callback != NULL) && transfer_complete) {
+		// Transfer complete, call user callback
+		data[channel].dma_callback(dev, data[channel].cb_data, channel + 1, 0);
+	}
+	// Enable IRQ
+	irq_enable(irq_number);
```

Comment:
Is it useful?

Reply:
Yes, we have an issue in the existing wiseconnect SDK which is fixed by this change.

Reply:
Sure, I will add the comment